client/asset/btc/btc.go

```diff
@@ -4276,9 +4310,83 @@ func (btc *baseWallet) DepositAddress() (string, error) {
 // RedemptionAddress gets an address for use in redeeming the counterparty's
 // swap. This would be included in their swap initialization.
 func (btc *baseWallet) RedemptionAddress() (string, error) {
+	return btc.recyclableAddress()
+}
+
+// A recyclable address is a redemption or refund address that may be recycled
+// if unused. If already recycled addresses are available, one will be returned.
+func (btc *baseWallet) recyclableAddress() (string, error) {
+	var recycledAddr string
+	btc.recycledAddrMtx.Lock()
+	nRecycles := len(btc.recycledAddrs)
+	for addr := range btc.recycledAddrs {
+		if owns, err := btc.OwnsDepositAddress(addr); owns {
+			delete(btc.recycledAddrs, addr)
+			recycledAddr = addr
+			break
+		} else if err != nil {
+			btc.log.Errorf("Error checking ownership of recycled address %q: %v", addr, err)
+			// Don't delete the address in case it's just a network error for
+			// an rpc wallet or something.
+		} else { // we don't own it
+			delete(btc.recycledAddrs, addr)
+		}
+	}
+	if len(btc.recycledAddrs) != nRecycles {
+		btc.writeRecycledAddrsToFile()
```

Comment:
Performing file system operations while a mutex is locked is not my favorite thing in the world.  Alright if workarounds are ugly or a lot of work, but one option is `func (btc *baseWallet) writeRecycledAddrsToFile(addrs []string)`, called after unlock.

A second idea is to defer writing of the file until the wallet is shutdown.  i.e. `(btc *baseWallet) connect(ctx context.Context)` could end with:

```go
	wg.Add(1)
	go func() {
		<-ctx.Done()
		btc.writeRecycledAddrsToFile()
	}()

	return &wg, nil
}
```

We don't seem to read from the file except on wallet start, so we only really need to write to it on shutdown.  The consequences of a stale file left from a crashed dexc are minimal.